core/src/main/java/org/pentaho/di/connections/vfs/DefaultVFSConnectionFileNameTransformer.java

```diff
@@ -0,0 +1,148 @@
+    // Provider URIs should not contain the name of the connection.
+    // Not even root uris, such as `s3://`. Should not become `s3://my-connection`!
+    // A folder with such a name, and in this case, given the hasBuckets condition, a bucket,
+    // will not be known by `KettleVFS` or the connection type's `VFSFileProvider`.
+    // if ( URI_ROOT.matcher( url ).matches() && details.hasBuckets() ) {
+    //   url += details.getName();
+    // }
+
+    return providerUriBuilder.toString();
+  }
+
+  // TODO: The connection root provider URI is being repeatedly determined, but is actually constant as long as the
+  // connection details does not change its configuration. Ideally, some kind of caching would be used.
+  // Only the "rest path" differs each time. The problem being there is no way to know if connection details have
+  // changed. These don't even implement hashCode, atm. Could use the connection manager to store and manage this
+  // cache, by name, invalidating it whenever a connection is loaded/saved. Would still need to deal with in-memory
+  // connections used by test(..) somehow.
+  protected void appendProviderUriConnectionRoot( @NonNull StringBuilder providerUriBuilder, @NonNull T details )
+    throws FileSystemException {
+
+    providerUriBuilder
+      .append( details.getType() )
+      .append( "://" );
+
+    // Examples: "hcp://" | "s3://"
+
+    appendProviderUriDomain( providerUriBuilder, details );
+
+    // Examples: "hcp://domain.my:443"
+
+    appendProviderUriRootPath( providerUriBuilder, details );
+
+    // Examples: "hcp://domain.my:443/root/path" | "local:///C:/root/path"
+  }
+
+  protected void appendProviderUriDomain( @NonNull StringBuilder providerUriBuilder, @NonNull T details )
+    throws FileSystemException {
+    // Assumed to not require any percent encoding. Limited to URI authority format / characters.
+    getConnectionFileNameUtils().appendPath( providerUriBuilder, details.getDomain() );
+  }
+
+  protected void appendProviderUriRootPath( @NonNull StringBuilder providerUriBuilder, @NonNull T details )
+    throws FileSystemException {
+
+    if ( details.isSupportsRootPath() ) {
+      // Root path is assumed to already be percent-encoded according to the rules of the specific
+      // file/connection provider.
+      getConnectionFileNameUtils().appendPath( providerUriBuilder, vfsConnectionManager.getResolvedRootPath( details ) );
+    }
+  }
+
+  protected void appendProviderUriRestPath( @NonNull StringBuilder providerUriBuilder,
+                                            @NonNull String restPath,
+                                            @NonNull T details )
+    throws FileSystemException {
+
+    getConnectionFileNameUtils().appendPath( providerUriBuilder, restPath );
+  }
+  // endregion
+
+  // region toPvfsFileName
+  @NonNull
+  @Override
+  public ConnectionFileName toPvfsFileName( @NonNull FileName providerFileName, @NonNull T details )
```

Comment:
Reuses the (possibly overridden) logic in `toProviderUri` to implement the inverse operation in a generic way, with no type-specific knowledge.